### PR TITLE
Add Light Unit Slider (Light Intensity, Temperature, Exposure)

### PR DESCRIPTION
**Pull request by johnpars**
### Purpose of this PR
This PR introduces the [proposed improvement](https://www.figma.com/file/Jb2OfbkPvAjlz9RbyMhnvf/HDRP-UI-IMPROVEMENTS?node-id=0%3A1) to UI/UX for light units. It adds a new slider UI drawer that loads a table of helpful value ranges, providing more context to the user about light unit values; through icons, markers, and tooltips. This will help the user to more easily reason about the variety of physically based values in HDRP, when authoring lighting and exposure.

## Light Intensity Units (EV, Nits, Lumen, Candela, Lux)
LightIntensity

## Temperature Units (Kelvin)
Temperature

## Exposure Units (EV)
Exposure

## Editor Theme
EditorThemes

## Slider is constant across all punctual light intensity units and spot reflector
ConstantSlider

---
### Testing status

- [ ] Checked new UI names with UX convention
- [x] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)

---
### Comments to reviewers

- This PR implements the new UI drawer for **Light Intensities**, **Temperature**, and **Exposure** (volume component), but has not yet been integrated to the **scene view override exposure**, or the **physically based sky exposure**.
- The slider for light intensities and exposure is modeled as a piece-wise function. At the cost of not being continuous and not monotonically increasing, it has the benefit of easily being able to accommodate the wide value distributions for certain light units, The PR does include an exponential version (which fits a curve to a low, mid, and high value), but its currently unused and can be removed. I have sent an email to our artists to gather additional feedback on this piece-wise distribution.
- For icon loading, currently they are not loaded with the HDRP editor resource loading mechanism (reloader). This is because we need to take special care to load icons so that they are skinned properly for the editor and it's color theme. Maybe it's possible to add Reloader support for icon case?

### CODE REVIEW

**Review by iM0ve — CHANGES_REQUESTED** [review]
**Review status:** In Progress

**Issues:**

1) **FIXED** - Can not input precise values. Same happens when upgrading a project, light values are changed.

values

2) **FIXED** - Overcast sky starts at 80k in the editor. In our own doc or wikipedia, we can see the value stated is 1000-2000 lux. Im not sure if this is an issue with terminology or with the values.

Screen Shot 2020-09-11 at 11 52 12

sources:
https://docs.unity3d.com/Packages/com.unity.render-pipelines.high-definition@10.0/manual/Physical-Light-Units.html
https://en.wikipedia.org/wiki/Daylight

3) Suggestion to change the tooltip for this one to something like "Exceeds realistic sunlight - 120K Lux". (no dot before dash). I think having static warning about 120k would benefit users more than just showing their own value in the tooltip, this way they can quickly evaluate how much they are overshooting.

Screen Shot 2020-09-11 at 11 59 34

4) **FIXED** - Concern. Im not sure about this one, but maybe the icons are a bit too transparent? I noticed the sun slider icon right away but somehow I didnt even see the Temperature one. Compared to all the other editor icons these are hard to spot. Matter of discussion.

Screen Shot 2020-09-11 at 12 22 59

5) Temperatures about sunlight are not too useful on Point/Spot/Area lights. We could show typical light bulb/candle values instead?

Screen Shot 2020-09-11 at 12 26 57

6) Undo no longer works. You can now also break rendering if you switch to EV100 and try to undo

break

7) I think the new Light Appearance mode might cause workflow issues. We might want to not implement the feature. [More details ](https://unity.slack.com/archives/C01A9B2JFNK/p1600248431067800)

8) When a value is converted to a different type, the icon indicates that it changes the lighting intensity. When in reality after conversion the value should be exactly the same.

bm1fqTEJZo


**Whats tested:**

1) Tooltips
2) Upgrading a project
3) UX
4) Functional testing

File: com.unity.render-pipelines.high-definition/Editor/Lighting/HDLightUI.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/HDLightUI.cs:62]: It should be `k_LightUnitSliderUIDrawer` instead

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:8]: A small comment here describing the purpose of the class would be nice here.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:19]: The descriptor is not meant to be modified, so you should add a `readonly`  keyword for safety.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:142]: This sounds like a useful utility here. Maybe you can extract a static function to convert a value to a readable human string (float value) -> string.
And then use it here.

Also, maybe we can use existing API like https://github.com/Humanizr/Humanizer#metric-numerals or similar?

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:163]: A short comment about the kind of remapping occurs here can be nice.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:171]: A small comment here describing the purpose of the class would be nice here.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:218]: A small comment here describing the purpose of the class would be nice here.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:158]: A small comment on protected API can be nice to know what is expected in the implementation.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:165]: A small comment on protected API can be nice to know what is expected in the implementation.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSliderSettings.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSliderSettings.cs:53]: `public static` don't follow the `k_` standard, you can directly use pascal case here and for other instances.

**Review by fredericv-unity3d — CHANGES_REQUESTED** [review]
I noticed a few minor things that needs changes and a few code standard issues.
But the overall PR is quite good.

Can you make the changes?

**Comment by johnpars** [comment]
Thanks @fredericv-unity3d - I have implemented almost all your notes barring the unit value humanizing. 

Note also that I am currently finishing an implementation of a piece-wise slider (due to the exponential slider currently not being fit well enough for the Lux value distribution). Additionally, I refactored the icon resource loading mechanism to support the different editor themes (HDRP's resource loading mechanism didn't make this easy). Due to those changes I will ping you for a re-review if that is alright.

@iM0ve - Note that at the time of your review the icon rendering indeed was not complete, but has since should now be correct for both editor themes. For the rest of your notes, I will investigate & follow up ASAP tomorrow.

**Comment by johnpars** [comment]
Hi @iM0ve, an update for you based on my recent commits:

> 1) Can not input precise values. Same happens when upgrading a project, light values are changed.

This should now be fixed. 

> 2) Overcast sky starts at 80k in the editor. In our own doc or wikipedia, we can see the value stated is 1000-2000 lux. Im not sure if this is an issue with terminology or with the values.

Awaiting input from UX on this one (cc @MaddalenaUnity). In the PR notes I linked to the original icon/value sheet which I derived all these values from.

> 3) Suggestion to change the tooltip for this one to something like "Exceeds realistic sunlight - 120K Lux". (no dot before dash).

Also waiting for UX thoughts on this one, I do agree though that it would be a positive change.

> 4) Concern. Im not sure about this one, but maybe the icons are a bit too transparent? I noticed the sun slider icon right away but somehow I didnt even see the Temperature one. Compared to all the other editor icons these are hard to spot. Matter of discussion.

Should now be fixed for both editor themes.

> 5) Temperatures about sunlight are not too useful on Point/Spot/Area lights. We could show typical light bulb/candle values instead?

Similar to 2, not fixed as I do not have the information/resources (icons) to improve these current values for punctual lights

> 6) Undo no longer works. You can now also break rendering if you switch to EV100 and try to undo

I believe this should now be fixed (both the issues with Undo as well as the NaN in EV units).

**Comment by johnpars** [comment]
Note: more discussion on this can be found on **#ux-lighting-units-improvements**

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:170]: This is mostly for your information, but you can use the [string interpolation syntax](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated) like this: `$"{baseTooltip}|{formatValue} {unit}"`

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:271]: `GetTransformation` is not exactly what is does.

I would use instead `NewTransformation` or `CreateTransformation` because it instantiate a lambda function (thus creating GC allocation).

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:308]: Note: swap can be done like this: `(x0, y0) = (y0, x0)`

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:310]: IMO swapping and then calling the method is a bit overkill here.

Just calling `GetTransformation(y0, y1, x0, x1)` is enough as the swapped values are not reused for something else.
I find it easier to understand the intent here.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:264]: As the algorithm does not change between transform and inverse transform, the use of `Func` is not appropriate: it is a reference function, possibly a null pointer.

You can store the `m` and `b` variables for direct and inverse transform and have a method to apply the algorithm instead.

If you are looking for reusable code you can have:
```
static float DoTransform(in float x, in float m, in float b) => (m * x) + b;
float Transform(in float x) => DoTransform(x, directM, directB);
float InverseTransform(in float x) => DoTransform(x, inverseM, inverseB);
```

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:269]: The key type `int` is disturbing here. You can directly use `Vector2` with the Dictionary and it will probably void a few bugs.

Dictionary will use the `GetHashCode()` method to bin the values and will use the `IEquatable<Self>.Equals(Self other)` method if `IEquatable<Self>` is implemented, otherwise it will use `object.Equals`.

So it means that if you use directly `Vector2` as Key type, you will achieve the same binning and still have `O(1)` time access, but it will handle equality cases when the `GetHashCode()` values are the same for different `Vector2`  (hash collision).

So other point is about understanding the code, `Vector2` is more explicit then `int` to indicate a range.

Can you please make this change?

Comment (fredericv-unity3d) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:269]: If I want to push further, I would use a new type pattern enclosing the range:
```
struct ValueRange: IEquatable<ValueRange>
{
    Vector2 m_Range;

    public ValueRange(in Vector2 value) => m_Range = value;
    public Vector2 ToVector2() => m_Range;
    // Insert generated Equality methods by ReSharper here
    
    public static explicit operator Vector2(in ValueRange v) => v.m_Range;
    public static explicit operator ValueRange(in Vector2 v) => new ValueRange(v);
}
```

Pros and cons of the newtype pattern:
Pros:
* The compiler will detect invalid uses of a Vector2 as a ValueRange and inversely. (Type check error)
* It is easier to review
* It helps a lot separating code in small understable and standalone pieces.

Cons:
* In C# it involves a lot of code.

I don't request the use of the newtype pattern as it is quite verbose, it is for your information you may need this pattern in the future.

(But I do request the change to `Vector2` at least :) )

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:334]: You should return a boolean to indicate if the search was successful.

**Review by fredericv-unity3d — CHANGES_REQUESTED** [review]
Thanks for resolving the issue and doing a clean PR 👍 .

I have still a few requests concerning the added code, can you make the changes?

(To be explicit: I also added potential changes at some points, but may be overkill for those use cases and I don't request to apply them, these are only for information)

**Review by iM0ve — CHANGES_REQUESTED** [review]
Updated status for existing issues. Added 7&8 
See my Review comment above

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (johnpars) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs]: Here I went with your other suggestion to just swap directly in the call

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (johnpars) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs]: Nice, thank you for all this info! Adding it to the toolbelt :) In this case, if it's alright, I have switched to using the `Vector2 `directly as the key. If I have extra time before end of week I may try to add this pattern, but currently need to allocate the remaining time to addressing the various QA and Artist request for the PR)

**Comment by johnpars** [comment]
Hi @iM0ve, thanks for the extra review. 

To address your notes:
- **3: Fixed**
- **7: Fixed** after discussion with artists in thread: the filter/temperaturewill be the default.
- **8: Fixed**, however there is an ongoing discussion with lighting artist on how this behavior should work when spot reflector is enabled.
- **5: Not Fixed**, this is a more UI/UX note which I will forward to @MaddalenaUnity

**Comment by johnpars** [comment]
Hi @fredericv-unity3d, thanks also for your review.

Your notes should all be addressed (note: for your note regarding Humanizing tooltip value, there is ongoing discussion with artists if it should be abbreviated or not).

If it's alright, I will request another re-review. I just pushed one last significant change to the slider that makes it always operate internally in Lumen. This was done to directly fix @iM0ve feedback number 8. 

It removed a lot of code (in the settings), added a slider specifically for converting between light units, and it elevated the access level of `ConvertLightIntensity` in `HDLightUI` to internal.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:274]: Now, the naming is still incorrect for other reason. `NewTransformation` was a good choice when it instantiated a lambda, now it would be more `ComputeTransformationParameters`.

I think that I had a confusing request in the previous review, the `NewTransformation` naming was good with the lambda pattern, but now we use directly an algorithm with parameters, the naming should be different.

Sorry about that.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:325]: Maybe indicate a small comment about the semantic of the returned boolean and that it performs a piece search based on `x` value.

It is not very obvious what kind of update is expected to be performed.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (fredericv-unity3d) [inline com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs:388]: `static readonly string[] k_UnitNames = `

**Review by fredericv-unity3d — CHANGES_REQUESTED** [review]
I found very minor issues that should be changed, can you update them?

**Comment by fredericv-unity3d** [comment]
You can request as many reviews as you need, it is always helpful!
Thanks for your work!

**Comment by iM0ve** [comment]
No new issues found. Updated the review. Only issue 5 still remains. Looks good. I will wait for all the changes to land and do one last check before approving.

File: com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs
Comment (johnpars) [reply com.unity.render-pipelines.high-definition/Editor/Lighting/LightUnit/LightUnitSlider.cs]: Hey @fredericv-unity3d, due to artist request we now just display the whole value here instead of abbreviating it into shorter terms. So I will mark resolved for now

**Comment by johnpars** [comment]
Thanks @fredericv-unity3d - I implemented your requests, and also added one last improvement that should satisfy the request of the artist (basically, need to satisfy the case of spot reflector; where the slider should be unchanged between light units). There was an unfortunate limitation to the `LightUtil` conversion that required me manually recompute some things (more info in the comment). Instead of improving the `LightUtil ` I decided to accomplish it all in the slider, as I was nervous about changing light unit conversion at this point since it might actually impact things on the runtime.

**Comment by johnpars** [comment]
Hi @iM0ve, for issue 5, it's certainly a doable change but I haven't been able to confirm alignment on it with UX. Additionally, I assume the change would also require new icons which I don't think are currently on hand from icon artist. If the change to the tooltip is enough however, it should be quickly doable.

Note also: I pushed a change to the light intensity slider that should be an improvement for how it behaves with spot reflector. Its probably the area of most concern in your next round of testing.

**Review by fredericv-unity3d — APPROVED** [review]
This is good for me!

**Comment by johnpars** [comment]
Note: This PR now fully implements the originally designed spec. However, we have the following feedback from artists:
- Improvement of iconography for temperature slider (specifically, waiting for two new icons - info is updated in the icon google sheet. Note that adding this should address the remaining issue flagged by @iM0ve).
- Add a right-click context menu on the slider, offering a dropdown of presets.

**Comment by sebastienlagarde** [comment]
Hey, so I will merge the current PR and we will do the improvement in a following one. THanks

**Review by sebastienlagarde — APPROVED** [review]

**Comment by iM0ve** [comment]
Still wanted to do a final pass before merge :/ please warn next time.
Did a quick check on reflector and Undo. Found that undo doesnt work with reflector properly and switching between certain light value types. For example works if Candela->Lumen->Undo and doesnt work Candela->Lux->Undo


type
refl

**Comment by sebastienlagarde** [comment]
arf sorry about that. was guessing it was done. Good catch

**Pull request merged**
